src/ObjectUtils/SymbolsFile.cpp

```diff
@@ -52,4 +55,45 @@ ErrorMessageOr<std::unique_ptr<SymbolsFile>> CreateSymbolsFile(
   return ErrorMessage{error_message};
 }
 
-}  // namespace orbit_object_utils
\ No newline at end of file
+// Demangling is quite expensive. We demangle in bulk here to easily measure the run time. It's also
+// a first step towards parallelization which could now be easily implemented.
+void DemangleSymbols(std::vector<FunctionSymbol>& function_symbols) {
+  ORBIT_SCOPE(absl::StrFormat("DemangleSymbols (%u)", function_symbols.size()).c_str());
+  for (FunctionSymbol& function_symbol : function_symbols) {
+    if (function_symbol.demangled_name.empty()) {
+      function_symbol.demangled_name =
+          absl::StrCat(llvm::demangle(function_symbol.name), function_symbol.argument_list);
+    }
+  }
+}
+
+// We centralize the creation of the ModuleSymbols proto, which is an expensive operation, so that
```

Comment:
Yes, this evolved and we used to extract symbols on the instance and send them, which we do not do anymore, so I think we could get rid of protos in this context. But I have not checked, and it might be that they are used (serialization) somehow.. maybe capture file, maybe presets. 

I would also be a bit hesitant to change everything, before we know how some currently moving parts are going to turn out. For auto symbol loading it is probably not relevant, but in other efforts it might be relevant that we are extracting symbols on the instance and send them somewhere. I am thinking about use cases where we have OrbitService running without a client attached ;)

Reply:
> I am not YET completely convinced. No we have an intermediate struct to be filled and moved around in addition to the proto creation. I see that is cheap, but its certainly not for free. And the benefit being that when we do adjustments to the proto creation, we can do it at this single point. Not sure if this is really better (while I totally see the point of deferring demangling to later).

@florian-kuebler , the creation of the raw struct is negligible compared to the cost of creating protos, and the strings are moved by the way, not copied. But really the goal here is to structure code so that it is extremely easy to parallelize. Both the proto creation and the demangling can now be trivially split on many worker threads. Moreover, as you know, this spawns from an effort to analyze and optimize the pdb loading time. In order to measure the cost of our different implementations, we need to isolate the different steps. It may seem like we're adding a little bit of cost now, but that will quickly go away with parallelization. Also, we now *reserve* the size of the protos, something we didn't do before, so for all we know it might already be faster in total. I could spend time measuring that, but I don't think it would be worth the effort, especially knowing that the entire reason for the change is to optimize symbol loading time. 

> What I am wondering right now, do we actually need protos here at all? @antonrohr correct me if I am wrong, but this is all happening on the client right. We will neither send the protos per grpc from the instance to the client nor save the symbols on disk (at least not yet). Can we maybe just remove the proto?

I really like the idea of getting rid of protos here if we can! As @antonrohr mentions, we might want to wait a little bit.

Reply:
> We centralize the creation of the ModuleSymbols proto, which is an expensive operation

I'm interested: I imagine that what makes filling the protos significantly slower than filling the structs here is that each `SymbolInfo` causes a new dynamic allocation in the `RepeatedPtrField` `symbol_infos` (while with a `std::vector` we allocate in place). Do I remember correctly? Does that benefit from parallelization?